### ISSUE TITLE
amp-fit-text: Content is unreadable when text style is adjusted

### ISSUE THREAD

**Issue by TetraLogicalHelpdesk** [ampproject/amphtml#29556]
## WCAG Level
Level AA

## Priority
High

## Type
Adaptability

## Pages/Components affected
* [amp-fit-text](https://amp.dev/documentation/components/amp-social-share/)

## Description
An important design consideration is to make sure that the display of text can be adjusted, so that line height, spacing between paragraphs, and spacing between words and letters can be changed without the content becoming unreadable.

Currently, if a user adjusts their letter/word/line spacing, all variations of `<amp-fit-text>` - depending on the specific size defined for the component and the amount of content inside them - can end up cropping / clipping content.

Below are screenshots from the [examples page](https://amp.dev/documentation/examples/components/amp-fit-text/?format=websites), after the spacing metrics have been forcibly changed via this [text spacing bookmarklet](https://www.html5accessibility.com/tests/tsbookmarklet.html). Note that in all instances, the content of some of the boxes is clipped (either at the start, or end, or both, of the example sentence).

### Basic usage
![Basic example with cropped content after spacing adjustment](https://user-images.githubusercontent.com/54933393/88787038-2e8eef00-d18b-11ea-8a1b-3140a8ce77dc.png)

### Aspect ratio
![Aspect ratio example with cropped content after spacing adjustment](https://user-images.githubusercontent.com/54933393/88787066-351d6680-d18b-11ea-9e87-e804ea1aa0e1.png)

### min-font-size
![min-font-size example with cropped content after spacing adjustment](https://user-images.githubusercontent.com/54933393/88787089-3c447480-d18b-11ea-8f58-6cc1b894dcd8.png)

### max-font-size
![max-font-size example with cropped content after spacing adjustment](https://user-images.githubusercontent.com/54933393/88787099-41a1bf00-d18b-11ea-9ebd-f45a75323c20.png)

## User impact
Someone with a cognitive or reading disability like Dyslexia may find content easier to read when there is more space between letters, words, lines and paragraphs. If content becomes hidden or obscured when they adjust the way text is displayed it makes the problem worse not better.

In this case, depending on the size of the `<amp-fit-text>` element and the amount of content, users may end up being unable to read/perceive the information contained in the element.

## Required solution
Make sure that the element takes spacing metrics into account when doing its fit calculations. As users may change these spacing metrics after the page has loaded/the elements have rendered, this may require listening for changes and redoing the calculation/fit dynamically.

## Test procedure(s)
Use these steps to confirm that the solution has been correctly applied:
1. Install a tool like the [ARC Toolkit](https://www.paciellogroup.com/toolkit/), which contains an option to change text spacing, or the [text spacing bookmarklet](https://www.html5accessibility.com/tests/tsbookmarklet.html).
2. Check that the text on the page responds to the changes.
3. Check that all content and functionality remains readable and available after the display of text has changed.

## Related standards
* [WCAG 2.1 Success Criterion 1.4.12 Text Spacing (Level AA)](https://www.w3.org/TR/WCAG21/#non-text-contrast)

## More information
* [W3C: Technique C35: Allowing for text spacing without wrapping](https://www.w3.org/WAI/WCAG21/Techniques/css/C35)
* [W3C: Technique C36: Allowing for text spacing override](https://www.w3.org/WAI/WCAG21/Techniques/css/C36)
* [Knowbility: Exploring WCAG 2.1: 1.4.12 Text Spacing](https://knowbility.org/blog/2018/WCAG21-1412TextSpacing/)
* [TPG: Short note on getting spaced out with SC 1.4.12 Text Spacing](https://developer.paciellogroup.com/blog/2018/05/short-note-on-getting-spaced-out-with-sc-1-4-12-text-spacing/)
* [TPG: ARC Toolkit](https://www.paciellogroup.com/toolkit/)

## Test data
Test date: July 2020

**Comment by rebeccanthomas** [comment]
Claiming this issue for the fixit. :)

**Comment by zhouyx** [comment]
Any update on this?

**Comment by samouri** [comment]
AMP doesn't react properly to many page events caused by external JS.
A ResizeObserver would be a natural way to solve this, but we don't polyfill that yet (its on the todo list).

I'm not sure of another way to solve this until ResOb is polyfilled :/
